<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | SQL: Projeto

Professora: [Mariane Neiva](https://www.linkedin.com/in/mariane-neiva/)<br>
Aluno: [Pedro Honorato](https://www.linkedin.com/in/pedro-honorato-6b01771b0/)<br>

[Acesse meu github](https://github.com/pedropauloha)

---

# **Sumário**

[**1. Dicionário de dados**](#1.-Dicionário-de-dados)  
[**2. EDA (Exploratory Data Analysis)**](#2.-EDA-(Exploratory-Data-Analysis))<br>
>[2.1 Conhecendo o dataset](#2.1-Conhecendo-o-dataset)<br>
>[2.2 Segundo etapa da analise - Distribuição](#2.2-Segundo-etapa-da-analise---Distribuição)<br>
>[2.3 Terceira etapa da analise - Agregações](#2.3-Terceira-etapa-da-analise---Agregações)<br>

[**3. Conclusões do material**](#3.-Conclusões-do-material)

In [1]:
# Importando pandas para auxiliar na visualização das bases de dados retornadas de cada query executada.
import pandas as pd

# 1. Dicionário de dados

O dataset usado neste projeto, para fins didáticos, é um recorte de cerca de 25% do dataset original. O conjunto de dados possui 13 colunas, sendo elas:

| Coluna                   | Descrição                                               | Tipo     |
| :----------------------- | :-----------------------------------------------------: | -------: |
| **idade**                | Idade do cliente                                        | *int*    | 
| **sexo**                 | Sexo do cliente (F ou M)                                | *string* |
| **dependentes**          | Número de dependentes do cliente                        | *int*    |
| **escolaridade**         | Nível de escolaridade do clientes                       | *string* |
| **estado_civil**         | Estado civil do cliente                                 | *string* |
| **salario_anual**        | Faixa salarial do cliente                               | *string* |
| **tipo_cartao**          | Tipo de cartao do cliente                               | *string* |
| **qtd_produtos**         | Quantidade de produtos comprados nos últimos 12 meses   | *bigint* |
| **iteracoes_12m**        | Quantidade de iterações/transacoes nos ultimos 12 meses | *int*    |
| **meses_inativo_12m**    | Quantidade de meses que o cliente ficou inativo         | *int*    |
| **limite_credito**       | Limite de credito do cliente                            | *float*  |
| **valor_transacoes_12m** | Valor das transações dos ultimos 12 meses               | *float*  |
| **qtd_transacoes_12m**   | Quantidade de transacoes dos ultimos 12 meses           | *int*    |

# Informações do projeto

O projeto foi criado como uma proposta de atividade final do curso de SQL da EBAC. Todo o processo de EDA foi feito com uso de linguagem SQL e foram criadas em ferramentas do ambiente da AWS, como: o [Athena](https://aws.amazon.com/pt/athena/) e [S3](https://aws.amazon.com/pt/s3/).


# 2. EDA (Exploratory Data Analysis)

## 2.1 Conhecendo o dataset

**Query 1** - Preview do Dataset com as 10 primeiras linhas
```sql
SELECT * FROM "default"."credito" limit 10;
```

In [2]:
dataset_preview = pd.read_csv("./queries/preview-dataset.csv",sep=',')
dataset_preview.head(10)

idade sexo  dependentes         escolaridade estado_civil   salario_anual  \
0     45    M            3         ensino medio       casado     $60K - $80K   
1     49    F            5             mestrado     solteiro  menos que $40K   
2     51    M            3             mestrado       casado    $80K - $120K   
3     40    F            4         ensino medio           na  menos que $40K   
4     40    M            3  sem educacao formal       casado     $60K - $80K   
5     44    M            2             mestrado       casado     $40K - $60K   
6     51    M            4                   na       casado         $120K +   
7     32    M            0         ensino medio           na     $60K - $80K   
8     37    M            3  sem educacao formal     solteiro     $60K - $80K   
9     48    M            2             mestrado     solteiro    $80K - $120K   

  tipo_cartao  qtd_produtos  iteracoes_12m  meses_inativo_12m  limite_credito  \
0        blue             5              3                  1        12691.51   
1        blue             6              2                  1         8256.96   
2        blue             4              0                  1         3418.56   
3        blue             3              1                  4         3313.03   
4        blue             5              0                  1         4716.22   
5        blue             3              2                  1         4010.69   
6        gold             6              3                  1        34516.72   
7      silver             2              2                  2        29081.49   
8        blue             5              0                  2        22352.50   
9        blue             6              3                  3        11656.41   

   valor_transacoes_12m  qtd_transacoes_12m  
0               1144.90                  42  
1               1291.45                  33  
2               1887.72                  20  
3               1171.56                  20  
4                816.08                  28  
5               1088.07                  24  
6               1330.87                  31  
7               1538.32                  36  
8               1350.14                  24  
9               1441.73                  32

**Query 2** - Tabela descritiva com todas as colunas do dataset
```sql
describe credito;
```

![describe](https://raw.githubusercontent.com/pedropauloha/ebac-ciencia-de-dados-sql/main/describe_db.png)

**Query 3** - Total de linhas do conjunto de dados
```sql
SELECT count(*) from credito;
```

In [3]:
count_rows = pd.read_csv("./queries/count_rows.csv",sep=',')
count_rows.head()

total_de_linhas
0             2564

**Query 4** - Valores possíveis na coluna Escolaridade
```sql
SELECT distinct escolaridade FROM credito;
```

In [4]:
distinct_escolaridade = pd.read_csv("./queries/distinct_escolaridade.csv",sep=',')
distinct_escolaridade.head()

escolaridade
0  ensino medio
1      mestrado
2     graduacao
3     doutorado
4            na

**Query 5** - Valores possíveis na coluna Estado Civil
```sql
SELECT distinct estado_civil FROM credito;
```

In [5]:
distinct_estado_civil = pd.read_csv("./queries/distinct_estado_civil.csv",sep=',')
distinct_estado_civil.head()

estado_civil
0   divorciado
1     solteiro
2       casado
3           na

**Informações básicas sobre o dataset**
<ol>
    <li>Total de 13 colunas</li>
    <li>2564 linhas</li>
    <li>Apresenta valores nulos</li>
    <li>Salario_anual é uma coluna categórica (string)</li>
    <li>Rastreio de informações dos últimos 12 meses</li>
</ol>


## 2.2 Segundo etapa da analise - Distribuição

### 2.2.1 Distribuição de clientes por gênero

**Query 6** - Distribuição de clientes por gênero
```sql
SELECT count(*) as contagem_genero,
    sexo
from credito
group by sexo
order by contagem_genero desc;
```

In [6]:
dist_sexo = pd.read_csv("./queries/dist_sexo.csv",sep=',')
dist_sexo.head()

contagem_genero sexo
0             1563    M
1             1001    F

### 2.2.2 Distribuição de clientes por categoria de salário

**Query 7** - Distribuição de clientes na base de dados pela categoria salarial
```sql
SELECT count(*) as contagem_de_clientes,
	salario_anual
from credito
group by salario_anual
order by CASE
    salario_anual
    WHEN 'menos que $40K' THEN 1
    WHEN '$40K - $60K' THEN 2
    WHEN '$60K - $80K' THEN 3
    WHEN '$80K - $120K' THEN 4
    WHEN '$120K +' THEN 5 ELSE 6
END;
```

In [7]:
count_clientes_salario = pd.read_csv("./queries/count_clientes_salario.csv",sep=',')
count_clientes_salario.head()

contagem_de_clientes   salario_anual
0                   701  menos que $40K
1                   467     $40K - $60K
2                   451     $60K - $80K
3                   488    $80K - $120K
4                   222         $120K +

### 2.2.3 Distribuição de clientes por tipo de cartão

**Query 8** - Distribuição de clientes na base de dados pelo tipo de cartão
```sql
select count(*) as contagem_de_clientes,
	tipo_cartao
from credito
group by tipo_cartao
order by contagem_de_clientes desc;
```

In [8]:
count_tipo_cartao = pd.read_csv("./queries/count_tipo_cartao.csv",sep=',')
count_tipo_cartao['pct'] = ((count_tipo_cartao['contagem_de_clientes'] / count_tipo_cartao['contagem_de_clientes'].sum()) *100)
count_tipo_cartao.head()

contagem_de_clientes tipo_cartao        pct
0                  2453        blue  95.670827
1                    93      silver   3.627145
2                    16        gold   0.624025
3                     2    platinum   0.078003

**Por que compreender a distribuição do dados do dataset?**

Compreender a distribuição da base de dados proporciona insights sobre os grupos de maior interesse para a empresa, seja para criar uma nova campanha específica para um público-alvo ou proporcionar um atendimento personalizado.

## 2.3 Terceira etapa da analise - Agregações

### 2.3.1 Agregações por idade

**Query 9** - Min/Max/Avg por idade - Análise da faixa etária dos clientes
```sql
select min(idade) as idade_minima,
	max(idade) as idade_maxima,
	round(avg(idade), 0) as idade_media
from credito;
```

In [9]:
min_max_avg_idade = pd.read_csv("./queries/min_max_avg_idade.csv",sep=',')
min_max_avg_idade.head()

idade_minima  idade_maxima  idade_media
0            26            73         46.0

### 2.3.2 Agregações por transações

**Query 10** - Min/Max/Avg por transações - Análise do consumo dos clientes
```sql
select min(valor_transacoes_12m) as min_transacoes,
	max(valor_transacoes_12m) as max_transacoes,
	round(avg(valor_transacoes_12m), 0) as avg_transacoes
from credito;
```

In [10]:
min_max_avg_transacoes = pd.read_csv("./queries/min_max_avg_transacoes.csv",sep=',')
min_max_avg_transacoes.head()

min_transacoes  max_transacoes  avg_transacoes
0          510.16         4776.58          1820.0

### 2.3.3 Top 10 clientes que possuem os maiores creditos e suas características por escolaridade, tipo do cartão e gênero

**Query 11** - Top 10 clientes que possuem os maiores creditos e suas características
```sql
select max(limite_credito) as limite_credito,
	escolaridade,
	tipo_cartao,
	sexo
from credito
where escolaridade != 'na'
	and tipo_cartao != 'na'
group by escolaridade,
	tipo_cartao,
	sexo
order by limite_credito desc
limit 10;
```

In [11]:
limite_max_esc_cart_sexo = pd.read_csv("./queries/limite_max_esc_cart_sexo.csv",sep=',')
limite_max_esc_cart_sexo.head(10)

limite_credito         escolaridade tipo_cartao sexo
0        34516.99  sem educacao formal        blue    M
1        34516.98             mestrado        gold    M
2        34516.97             mestrado        blue    M
3        34516.96  sem educacao formal      silver    M
4        34516.96            doutorado    platinum    F
5        34516.95         ensino medio        gold    M
6        34516.94            graduacao        gold    M
7        34516.94            graduacao        gold    F
8        34516.89            graduacao      silver    M
9        34516.74         ensino medio        blue    M

### 2.3.4 Características dos clientes com as 10 maiores médias de valores gastos nos últimos 12 meses

**Query 12** - Características dos clientes com as 10 maiores médias de valores gastos nos últimos 12 meses
```sql
select avg(valor_transacoes_12m) as total_transacoes,
	escolaridade,
	sexo,
	estado_civil,
	tipo_cartao
from credito
where escolaridade != 'na'
	and estado_civil != 'na'
	and tipo_cartao != 'na'
group by escolaridade,
	sexo,
	estado_civil,
	tipo_cartao
order by total_transacoes desc
limit 10;
```

In [12]:
media_valor_gasto = pd.read_csv("./queries/media_valor_gasto.csv",sep=',')
media_valor_gasto.head(10)

total_transacoes         escolaridade sexo estado_civil tipo_cartao
0         4149.1800            doutorado    M       casado        gold
1         3899.2700             mestrado    M     solteiro        gold
2         3360.6000             mestrado    F   divorciado      silver
3         3205.1200             mestrado    M       casado        gold
4         2897.4534             mestrado    F       casado      silver
5         2574.8000            graduacao    F     solteiro        gold
6         2535.8076            doutorado    F   divorciado        blue
7         2500.9550             mestrado    F     solteiro      silver
8         2407.4475  sem educacao formal    F     solteiro      silver
9         2327.1100             mestrado    F     solteiro        gold

### 2.3.5 Valores gastos nos últimos 12 meses por gênero

**Query 13** - Valores gastos nos últimos 12 meses por gênero
```sql
select max(valor_transacoes_12m) as maximo_valor_gasto,
	avg(valor_transacoes_12m) as media_valor_gasto,
	min(valor_transacoes_12m) as minimo_valor_gasto,
	sexo
from credito
group by sexo;
```

In [13]:
gasto_por_genero = pd.read_csv("./queries/gasto_por_genero.csv",sep=',')
gasto_por_genero.head(10)

maximo_valor_gasto  media_valor_gasto  minimo_valor_gasto sexo
0             4776.58          1839.6226              510.16    F
1             4686.93          1807.9680              530.36    M

# 3. Conclusões do material

**Insights extraídos da análise**

<ol>
    <li>A maior parte dos clientes possui renda até 40K;</li>
    <li>A maior parte dos clientes é homem;</li>
    <li>96% dos clientes possuem cartão "blue";</li>
    <li>A média de idade dos clientes é de 46 anos;</li>
    <li>O valor gasto por todos os clientes é de 1820;</li>
    <li>Escolaridade mais elevadas e tipos de cartões mais exclusivos não parecem influenciar totalmente no máximo de limite que um cliente pode ter disponível;</li>
    <li>Os dados indicam que mulheres, em média, gastam mais que homens;</li>
    <li>Os dados indicam que níveis de escolaridade mais elevados somam gastos mais altos.</li>
</ol>
